### PR TITLE
RWA-1593: Code changes to fix task termination error handling

### PR DESCRIPTION
**Pull request by rasikasri**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-1593

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserIdCacheTest.java [hunk 1]
```diff
@@ -27,7 +28,7 @@ public class IdamServiceUserIdCacheTest {
     @Autowired
     private IdamService idamService;
 
-    @Test
+//    @Test
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserIdCacheTest.java:31]: did you mean to comment out?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserInfoCacheTest.java [hunk 2]
```diff
@@ -27,7 +28,7 @@ public class IdamServiceUserInfoCacheTest {
     @Autowired
     private IdamService idamService;
 
-    @Test
+//    @Test
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserInfoCacheTest.java:31]: same here

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java [hunk 3]
```diff
@@ -654,15 +654,27 @@ public long getTaskCount(SearchTaskRequest searchTaskRequest) {
...
-        task.setState(CFTTaskState.TERMINATED);
-        task.setTerminationReason(terminateInfo.getTerminateReason());
-        //Perform Camunda updates
-        camundaService.deleteCftTaskState(taskId);
-        //Commit transaction
-        cftTaskDatabaseService.saveTask(task);
+        TaskResource task = null;
+        try {
+            //Find and Lock Task
+            task = findByIdAndObtainLock(taskId);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java:660]: So does the transaction we open here close when we return?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java [hunk 4]
```diff
@@ -654,15 +654,27 @@ public long getTaskCount(SearchTaskRequest searchTaskRequest) {
...
+        } catch (ResourceNotFoundException e) {
+            //Perform Camunda updates
+            log.info("Task for id {} not found in the database, trying delete the task in camunda if exist", taskId);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java:663]: maybe a `warn` message?

**Review by raghera — COMMENTED** [review]
Looks good in general.  Just a couple of questions.

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserIdCacheTest.java (hunk 1 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/auth/IdamServiceUserIdCacheTest.java:31]: Nope, sorry. I'll remove it

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java (hunk 3 above)
Comment (rasikasri) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java:660]: Yes, and as it is read only call, doesn't make any difference either we rollback or commit

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java (hunk 4 above)
Comment (rasikasri) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java:663]: done

**Review by dharmendrak — APPROVED** [review]

**Pull request merged**